NetTopologySuite.IO.PostGis/PostGisWriter.cs

```diff
@@ -170,20 +149,20 @@ private static void Write(ICoordinateSequence sequence, Ordinates ordinates, Bin
                 if ((ordinates & Ordinates.Z) != 0)
                 {
                     double z = sequence.GetOrdinate(i, Ordinate.Z);
-                    if (double.IsNaN(z)) z = 0d;
+                    Debug.Assert(!double.IsNaN(z));
```

Comment:
We should just skip these two `Debug.Assert` lines (this one, for Z, and the other one, for M).  Nobody's ever going to see it since we don't release debug builds, and it can trigger even if the rest of our code is correct (since it depends fully on the input data).

This is assuming that I'm not wrong in my understanding that `Debug.Assert` calls are meant to inform the developer that, if the rest of our code is correct, its condition will always be true at the point in time when the `Debug.Assert` call is made.

Especially for this one (Z), I expect to see some `NaN` values on a somewhat regular basis, in cases where someone just creates a new geometry with all the default services that are set up for you (at least as of NTS 1.15... I'm still not sure how 1.16 will look), since I'm pretty sure that a 3-dimensional coordinate sequence is still the default even though NTS is advertised as a library for handling primarily 2-dimensional geometry.

Reply:
If I'm wrong and these `Debug.Assert` lines do in fact fulfill a purpose, then please use an overload such as [`Debug.Assert(bool, string)`](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.debug.assert?view=netframework-4.7.2#System_Diagnostics_Debug_Assert_System_Boolean_System_String_) that allows you to provide a message explaining in more detail what happened and, perhaps, why it's bad.

Reply:
Do you mean to say that it's possible to encounter a point `p` which has `p.CoordinateSequence.Ordinates` equal to `XYZ`, but when I access its Z coordinate I get a NaN (unset or something)? If so then the current code definitely isn't right and needs to be fixed - but I'm really not sure if we can actually send anything to PostgreSQL that makes sense... I don't think it'd accept NaN (but can check), so probably raising an exception is the way to go.

Let me know...

Reply:
> Do you mean to say that it's possible to encounter a point `p` which has `p.CoordinateSequence.Ordinates` equal to `XYZ`, but when I access its Z coordinate I get a NaN (unset or something)?

Yes, and it doesn't require anything esoteric to get into that situation either; someone who is unfamiliar with the idiosyncrasies of JTS / NTS will more likely than not stumble into it:
# ConsoleApp0.csproj
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp2.2</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="NetTopologySuite.Core" Version="1.15.1" />
  </ItemGroup>

</Project>
```

# Program.cs
```csharp
using System;
using GeoAPI.Geometries;
using NetTopologySuite.Geometries;

class Program
{
    static void Main()
    {
        var seq = new Point(0, 0).CoordinateSequence;
        Console.WriteLine(seq.Ordinates); // writes XYZ
        var z = seq.GetOrdinate(0, Ordinate.Z);
        Console.WriteLine(z); // writes NaN
    }
}
```

> I'm really not sure if we can actually send anything to PostgreSQL that makes sense... I don't think it'd accept NaN (but can check), so probably raising an exception is the way to go.

I don't have enough experience with the PostgreSQL / PostGIS side of things to be able to comment on that, but I suspect that this is a reason why the current version of the code fills in zero wherever it reads a NaN.

Reply:
> I don't have enough experience with the PostgreSQL / PostGIS side of things to be able to comment on that, but I suspect that this is a reason why the current version of the code fills in zero wherever it reads a NaN.

In PostGis, it is certainly possible to create a geometry that has NaN co-ordinate values. For example, setting the Z co-ord to NaN:

    ST_MakePoint(0, 0, 'NaN')

yields:

     010100008000000000000000000000000000000000000000000000F8FF

which suggests that PostGis uses `000000000000F8FF` (i.e. the standard IEEE 754 NaN) to represent NaN co-ordinate values. So it should be fine? Possibly no need even for the `Assert.Debug` statement?

Reply:
Thanks for testing that @peetw, agreed! I've removed the assertions, NaNs will now be written to PostgreSQL just like any other value.